main/src/ca/uwaterloo/flix/language/phase/Reducer.scala

```diff
@@ -33,131 +33,145 @@ import scala.jdk.CollectionConverters._
   * 2. Collect a set of all anonymous class / new object expressions
   * 3. Collect a flat set of all types of the program, i.e., if `List[String]` is
   *   in the list, so is `String`.
-  *
-  * This is actually achieved as two pseudo phases, first 1. and 2. and then 3.
-  * operating on a new root.
   */
 object Reducer {
 
   def run(root: Root)(implicit flix: Flix): Root = flix.phase("Reducer") {
-    implicit val ctx: SharedContext = SharedContext(new ConcurrentLinkedQueue)
+    implicit val ctx: SharedContext = SharedContext(new ConcurrentLinkedQueue, ConcurrentHashMap.newKeySet())
 
     val newDefs = ParOps.parMapValues(root.defs)(visitDef)
+    val defTypes = ctx.defTypes.asScala.toSet
 
-    val newRoot = root.copy(defs = newDefs, anonClasses = ctx.anonClasses.asScala.toList)
-    val types = typesOf(newRoot)
-    newRoot.copy(types = types)
+    val effectTypes = root.effects.values.toSet.flatMap(typesOfEffect)
```

Comment:
.values.toSet and flatMap can be expensive here (?)

Reply:
Its on effects so i guess its small, but we could traverse it in parallel like with defs. Ill add it to the followup

Reply:
yeah it can wait i think